// zarm/lib/NoticeBar/NoticeBar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Message = require('../Message');

var _Message2 = _interopRequireDefault(_Message);

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NoticeBar = function (_PureComponent) {
  _inherits(NoticeBar, _PureComponent);

  function NoticeBar(props) {
    _classCallCheck(this, NoticeBar);

    var _this = _possibleConstructorReturn(this, (NoticeBar.__proto__ || Object.getPrototypeOf(NoticeBar)).call(this, props));

    _this.state = {
      offset: 0
    };
    _this.moveInterval = null;
    return _this;
  }

  _createClass(NoticeBar, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      var autoscroll = this.props.autoscroll;

      if (!autoscroll) return;

      var distance = this.wrapper.offsetWidth - this.content.offsetWidth;
      if (distance > 0) return;

      var delay = 1000;
      this.moveInterval = setInterval(function () {
        var offset = _this2.state.offset;

        if ((offset < distance || offset >= 0) && delay > 0) {
          delay -= 50;
          return;
        }

        delay = 1000;
        offset = offset < distance ? 0 : offset - 1;

        _this2.setState({ offset: offset });
      }, 50);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.moveInterval) {
        clearInterval(this.moveInterval);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          children = _props.children,
          autoscroll = _props.autoscroll,
          others = _objectWithoutProperties(_props, ['prefixCls', 'children', 'autoscroll']);

      return _react2.default.createElement(
        _Message2.default,
        _extends({}, others, { size: 'lg' }),
        _react2.default.createElement(
          'div',
          { className: prefixCls, ref: function ref(ele) {
              _this3.wrapper = ele;
            } },
          _react2.default.createElement(
            'div',
            { className: prefixCls + '-body', ref: function ref(ele) {
                _this3.content = ele;
              }, style: { left: this.state.offset } },
            children
          )
        )
      );
    }
  }]);

  return NoticeBar;
}(_react.PureComponent);

NoticeBar.propTypes = {
  prefixCls: _propTypes2.default.string,
  className: _propTypes2.default.string,
  theme: _Message2.default.propTypes.theme,
  icon: _Message2.default.propTypes.icon,
  hasArrow: _Message2.default.propTypes.hasArrow,
  hasClosable: _Message2.default.propTypes.hasClosable,
  autoscroll: _propTypes2.default.bool
};

NoticeBar.defaultProps = {
  prefixCls: 'za-noticebar',
  theme: 'warning',
  icon: _react2.default.createElement(_Icon2.default, { type: 'broadcast' }),
  hasArrow: false,
  hasClosable: false,
  autoscroll: false
};

exports.default = NoticeBar;
