// zarm/lib/Toast/Toast.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _Mask = require('../Mask');

var _Mask2 = _interopRequireDefault(_Mask);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Toast = function (_PureComponent) {
  _inherits(Toast, _PureComponent);

  function Toast(props) {
    _classCallCheck(this, Toast);

    var _this = _possibleConstructorReturn(this, (Toast.__proto__ || Object.getPrototypeOf(Toast)).call(this, props));

    _this.enter = function (_ref) {
      var stayTime = _ref.stayTime,
          onMaskClick = _ref.onMaskClick;

      _this.setState({
        isShow: true
      });

      if (stayTime === 0) return;

      _this.timer = setTimeout(function () {
        typeof onMaskClick === 'function' && onMaskClick();
        clearTimeout(_this.timer);
      }, stayTime);
    };

    _this.leave = function () {
      _this.setState({
        isShow: false
      });
    };

    _this.timer = null;
    _this.state = {
      isShow: false
    };
    return _this;
  }

  _createClass(Toast, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.props.visible) {
        this.enter(this.props);
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      clearTimeout(this.timer);

      if (nextProps.visible) {
        this.enter(nextProps);
      } else {
        this.leave();
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      clearTimeout(this.timer);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          prefixCls = _props.prefixCls,
          className = _props.className,
          visible = _props.visible,
          children = _props.children,
          onMaskClick = _props.onMaskClick;


      var cls = (0, _classnames3.default)('' + prefixCls, className, _defineProperty({}, prefixCls + '-open', this.state.isShow));

      return _react2.default.createElement(
        'div',
        { className: cls },
        _react2.default.createElement(
          'div',
          { className: prefixCls + '-container' },
          children
        ),
        _react2.default.createElement(_Mask2.default, { visible: visible, type: 'transparent', onClose: onMaskClick })
      );
    }
  }]);

  return Toast;
}(_react.PureComponent);

Toast.propTypes = {
  prefixCls: _propTypes2.default.string,
  className: _propTypes2.default.string,
  visible: _propTypes2.default.bool,
  stayTime: _propTypes2.default.number, // eslint-disable-line
  onMaskClick: _Mask2.default.propTypes.onClose
};

Toast.defaultProps = {
  prefixCls: 'za-toast',
  visible: false,
  stayTime: 3000
};

exports.default = Toast;
