// zarm/lib/Progress/Progress.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _Spinner = require('../Spinner');

var _Spinner2 = _interopRequireDefault(_Spinner);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Progress = function (_PureComponent) {
  _inherits(Progress, _PureComponent);

  function Progress() {
    _classCallCheck(this, Progress);

    return _possibleConstructorReturn(this, (Progress.__proto__ || Object.getPrototypeOf(Progress)).apply(this, arguments));
  }

  _createClass(Progress, [{
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          className = _props.className,
          theme = _props.theme,
          shape = _props.shape,
          strokeWidth = _props.strokeWidth,
          percent = _props.percent,
          children = _props.children,
          others = _objectWithoutProperties(_props, ['prefixCls', 'className', 'theme', 'shape', 'strokeWidth', 'percent', 'children']);

      var cls = (0, _classnames3.default)('' + prefixCls, className, (_classnames = {}, _defineProperty(_classnames, 'theme-' + theme, !!theme), _defineProperty(_classnames, 'shape-' + shape, !!shape), _classnames));

      var innerRender = shape === 'circle' ? _react2.default.createElement(
        'div',
        { className: prefixCls + '-inner' },
        _react2.default.createElement(_Spinner2.default, { theme: theme, strokeWidth: strokeWidth, percent: percent })
      ) : _react2.default.createElement(
        'div',
        { className: prefixCls + '-inner', style: { height: strokeWidth } },
        _react2.default.createElement('div', { className: prefixCls + '-bg', style: { width: percent + '%' } })
      );

      return _react2.default.createElement(
        'div',
        _extends({ className: cls }, others),
        innerRender,
        children && _react2.default.createElement(
          'div',
          { className: prefixCls + '-text' },
          children
        )
      );
    }
  }]);

  return Progress;
}(_react.PureComponent);

Progress.propTypes = {
  prefixCls: _propTypes2.default.string,
  className: _propTypes2.default.string,
  theme: _Spinner2.default.propTypes.theme,
  shape: _propTypes2.default.oneOf(['line', 'circle']),
  strokeWidth: _Spinner2.default.propTypes.strokeWidth,
  percent: _Spinner2.default.propTypes.percent

};

Progress.defaultProps = {
  prefixCls: 'za-progress',
  theme: _Spinner2.default.defaultProps.theme,
  shape: 'line',
  strokeWidth: _Spinner2.default.defaultProps.strokeWidth,
  percent: _Spinner2.default.defaultProps.percent
};

exports.default = Progress;
