// zarm/lib/Tab/Tab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _TabPanel = require('./TabPanel');

var _TabPanel2 = _interopRequireDefault(_TabPanel);

var _Swipe = require('../Swipe');

var _Swipe2 = _interopRequireDefault(_Swipe);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function getSelectIndex(children) {
  var selectIndex = void 0;
  _react2.default.Children.forEach(children, function (item, $index) {
    if (item.props && item.props.selected) {
      selectIndex = $index;
    }
  });
  return selectIndex;
}

var Tab = function (_PureComponent) {
  _inherits(Tab, _PureComponent);

  function Tab(props) {
    _classCallCheck(this, Tab);

    var _this = _possibleConstructorReturn(this, (Tab.__proto__ || Object.getPrototypeOf(Tab)).call(this, props));

    _this.state = {
      value: props.value || props.defaultValue || getSelectIndex(props.children) || 0
    };
    return _this;
  }

  _createClass(Tab, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if ('value' in nextProps || getSelectIndex(nextProps.children)) {
        this.setState({
          value: nextProps.value || nextProps.defaultValue || getSelectIndex(nextProps.children) || 0
        });
        typeof nextProps.onChange === 'function' && nextProps.onChange(nextProps.value);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          className = _props.className,
          theme = _props.theme,
          lineWidth = _props.lineWidth,
          disabled = _props.disabled,
          canSwipe = _props.canSwipe,
          children = _props.children,
          _onChange = _props.onChange;


      var classes = (0, _classnames3.default)('' + prefixCls, className, _defineProperty({}, 'theme-' + theme, !!theme));

      // 渲染选项
      var tabsRender = _react2.default.Children.map(children, function (item, $index) {
        var itemCls = (0, _classnames3.default)(prefixCls + '-header-item', item.props.className, {
          disabled: disabled || item.props.disabled,
          active: _this2.state.value === $index
          // hasline,
        });

        return _react2.default.createElement(
          'li',
          {
            role: 'tab',
            key: $index,
            className: itemCls,
            onClick: function onClick() {
              if (disabled || item.props.disabled) return;
              _this2.setState({ value: $index });
              typeof _onChange === 'function' && _onChange($index);
              canSwipe && _this2.swipe.onSlideTo($index);
            } },
          item.props.title
        );
      });

      // 渲染内容
      var contentRender = void 0;

      if (canSwipe) {
        contentRender = _react2.default.createElement(
          _Swipe2.default,
          {
            showPagination: false,
            activeIndex: this.state.value,
            ref: function ref(ele) {
              _this2.swipe = ele;
            },
            onChange: function onChange(value) {
              _this2.setState({ value: value });
              typeof _onChange === 'function' && _onChange(value);
            } },
          _react2.default.Children.map(children, function (item) {
            return _react2.default.createElement(
              'div',
              null,
              item.props.children
            );
          })
        );
      } else {
        contentRender = _react2.default.Children.map(children, function (item, $index) {
          return _react2.default.createElement(_TabPanel2.default, _extends({}, item.props, { selected: _this2.state.value === $index }));
        });
      }

      var lineStyle = {
        width: 100 / children.length + '%',
        left: this.state.value / children.length * 100 + '%'
        // right: `${(children.length - this.state.value - 1) / children.length * 100}%`,
        // transition: `right 0.3s cubic-bezier(0.35, 0, 0.25, 1), left 0.3s cubic-bezier(0.35, 0, 0.25, 1) 0.09s`,
      };

      var lineInnerRender = void 0;
      if (lineWidth) {
        var w = void 0;
        if (lineWidth === 'auto') {
          w = lineWidth;
        } else {
          w = lineWidth;
        }
        lineStyle.backgroundColor = 'transparent';
        lineInnerRender = _react2.default.createElement('span', { className: prefixCls + '-line-inner', style: { width: w } });
      }

      return _react2.default.createElement(
        'div',
        { className: classes },
        _react2.default.createElement(
          'div',
          { className: prefixCls + '-header' },
          _react2.default.createElement(
            'ul',
            { role: 'tablist' },
            tabsRender
          ),
          _react2.default.createElement(
            'div',
            { className: prefixCls + '-line', style: lineStyle },
            lineInnerRender
          )
        ),
        _react2.default.createElement(
          'div',
          { className: prefixCls + '-container' },
          contentRender
        )
      );
    }
  }]);

  return Tab;
}(_react.PureComponent);

Tab.propTypes = {
  prefixCls: _propTypes2.default.string,
  className: _propTypes2.default.string,
  theme: _propTypes2.default.oneOf(['default', 'primary', 'info', 'success', 'warning', 'error']),
  lineWidth: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
  disabled: _propTypes2.default.bool,
  canSwipe: _propTypes2.default.bool,
  onChange: _propTypes2.default.func
};

Tab.defaultProps = {
  prefixCls: 'za-tab',
  theme: 'primary',
  disabled: false,
  canSwipe: false
};

exports.default = Tab;
