// zarm/lib/Input/InputTextarea.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _autosize = require('autosize');

var _autosize2 = _interopRequireDefault(_autosize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var InputTextarea = function (_PureComponent) {
  _inherits(InputTextarea, _PureComponent);

  function InputTextarea(props) {
    _classCallCheck(this, InputTextarea);

    var _this = _possibleConstructorReturn(this, (InputTextarea.__proto__ || Object.getPrototypeOf(InputTextarea)).call(this, props));

    _this.onInputChange = function (e) {
      var onChange = _this.props.onChange;

      var length = e.target.value.length;
      _this.setState({ length: length });
      typeof onChange === 'function' && onChange(e);
    };

    _this.initAutosize = function () {
      var autosize = _this.props.autosize;

      autosize && (0, _autosize2.default)(_this.input);
    };

    _this.updateAutosize = function (prevProps) {
      if (prevProps.style !== _this.props.style || prevProps.className !== _this.props.className) {
        _autosize2.default.update(_this.input);
      }
    };

    _this.destroyAutosize = function () {
      var autosize = _this.props.autosize;

      autosize && _autosize2.default.destroy(_this.input);
    };

    _this.state = {
      length: (props.value || props.defaultValue || '').length
    };
    return _this;
  }

  _createClass(InputTextarea, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.initAutosize();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      this.updateAutosize(prevProps);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.destroyAutosize();
    }

    // 初始化自适应高度


    // 销毁自适应高度

  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          className = _props.className,
          maxLength = _props.maxLength,
          disabled = _props.disabled,
          autosize = _props.autosize,
          showLength = _props.showLength,
          others = _objectWithoutProperties(_props, ['prefixCls', 'className', 'maxLength', 'disabled', 'autosize', 'showLength']);

      var cls = (0, _classnames2.default)(prefixCls, prefixCls + '-textarea', className, disabled);
      var textLengthRender = showLength && maxLength && _react2.default.createElement(
        'div',
        { className: prefixCls + '-length' },
        this.state.length + '/' + maxLength
      );

      return _react2.default.createElement(
        'div',
        { className: cls },
        _react2.default.createElement('textarea', _extends({}, others, {
          ref: function ref(ele) {
            _this2.input = ele;
          },
          disabled: disabled,
          maxLength: maxLength,
          onChange: this.onInputChange
        })),
        textLengthRender
      );
    }
  }]);

  return InputTextarea;
}(_react.PureComponent);

InputTextarea.propTypes = {
  prefixCls: _propTypes2.default.string,
  className: _propTypes2.default.string,
  value: _propTypes2.default.string,
  defaultValue: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  rows: _propTypes2.default.number,
  autosize: _propTypes2.default.bool,
  maxLength: _propTypes2.default.number,
  showLength: _propTypes2.default.bool,
  onChange: _propTypes2.default.func
};

InputTextarea.defaultProps = {
  prefixCls: 'za-input',
  disabled: false,
  autosize: false,
  showLength: false
};

exports.default = InputTextarea;
