// zarm/lib/Cell/Cell.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Cell = function (_PureComponent) {
  _inherits(Cell, _PureComponent);

  function Cell() {
    _classCallCheck(this, Cell);

    return _possibleConstructorReturn(this, (Cell.__proto__ || Object.getPrototypeOf(Cell)).apply(this, arguments));
  }

  _createClass(Cell, [{
    key: 'render',
    value: function render() {
      var _classnames;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          className = _props.className,
          theme = _props.theme,
          hasArrow = _props.hasArrow,
          icon = _props.icon,
          title = _props.title,
          description = _props.description,
          help = _props.help,
          disabled = _props.disabled,
          onClick = _props.onClick,
          children = _props.children,
          others = _objectWithoutProperties(_props, ['prefixCls', 'className', 'theme', 'hasArrow', 'icon', 'title', 'description', 'help', 'disabled', 'onClick', 'children']);

      var cls = (0, _classnames3.default)((_classnames = {}, _defineProperty(_classnames, '' + prefixCls, true), _defineProperty(_classnames, className, !!className), _defineProperty(_classnames, 'theme-' + theme, !!theme), _defineProperty(_classnames, 'disabled', disabled), _defineProperty(_classnames, 'is-link', !disabled && !!onClick), _defineProperty(_classnames, 'has-icon', !!icon), _defineProperty(_classnames, 'has-arrow', hasArrow), _classnames));

      var iconRender = icon ? _react2.default.createElement(
        'div',
        { className: prefixCls + '-icon' },
        icon
      ) : null;

      var titleRender = title ? _react2.default.createElement(
        'div',
        { className: prefixCls + '-title' },
        title
      ) : null;

      var contentRender = children ? _react2.default.createElement(
        'div',
        { className: prefixCls + '-content' },
        children
      ) : null;

      var arrowRender = hasArrow ? _react2.default.createElement('div', { className: prefixCls + '-arrow' }) : null;

      var helpRender = help ? _react2.default.createElement(
        'div',
        { className: prefixCls + '-help' },
        help
      ) : null;

      return _react2.default.createElement(
        'div',
        _extends({ className: cls, onClick: onClick, onTouchStart: function onTouchStart() {} }, others),
        _react2.default.createElement(
          'div',
          { className: prefixCls + '-inner' },
          _react2.default.createElement(
            'div',
            { className: prefixCls + '-header' },
            iconRender
          ),
          _react2.default.createElement(
            'div',
            { className: prefixCls + '-body' },
            titleRender,
            contentRender
          ),
          _react2.default.createElement(
            'div',
            { className: prefixCls + '-footer' },
            description
          ),
          arrowRender
        ),
        helpRender
      );
    }
  }]);

  return Cell;
}(_react.PureComponent);

Cell.propTypes = {
  prefixCls: _propTypes2.default.string,
  className: _propTypes2.default.string,
  hasArrow: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool
};

Cell.defaultProps = {
  prefixCls: 'za-cell',
  hasArrow: false,
  disabled: false
};

exports.default = Cell;
